src/Stack/PackageIndex.hs

```diff
@@ -365,6 +378,23 @@ lookupPackageVersions :: PackageName -> Map PackageIdentifier a -> Set Version
 lookupPackageVersions pkgName pkgCaches =
     Set.fromList [v | PackageIdentifier n v <- Map.keys pkgCaches, n == pkgName]
 
+-- | Access the package caches from 'IO'.
+withPackageCaches
+    :: (MonadIO m, MonadReader env m, HasConfig env, MonadLogger m, HasHttpManager env, MonadBaseControl IO m, MonadCatch m)
+    => (IO (Map PackageIdentifier (PackageIndex, PackageCache)) -> m a)
+    -> m a
+withPackageCaches inner = do
+    runInBase <- liftBaseWith $ \run -> return (void . run)
+    let getPackageCaches' :: IO (Map PackageIdentifier (PackageIndex, PackageCache))
+        getPackageCaches' = do
+            -- TODO: Use monad base control properly.
+            icaches <- newIORef (error "Impossible evaluation in withPackageCaches")
+            runInBase $ do
+                caches <- getPackageCaches
+                liftIO $ writeIORef icaches caches
+            readIORef icaches
+    inner getPackageCaches'
```

Comment:
I've been looking how to do this properly with `MonadBaseControl` but failed. My impression was that you'd always have to evaluate `getPackageCaches` before you could pass it to `inner`.